File: r-inla/validate_spde.R
# SPDE mesh / FEM / field demo using rust-inla only (no classic INLA package).
#
# Mirrors the usual R-INLA workflow:
#   coords → mesh → Matérn SPDE FEM (C, G) → Q → sample field → figures
#
# Usage (from r-inla/):
#   Rscript validate_spde.R
# or via validate_spde.sh (builds the shared library first).

args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)
script_dir <- if (length(file_arg)) {
  dirname(normalizePath(sub("^--file=", "", file_arg[1])))
} else {
  getwd()
}
setwd(script_dir)

source("R/inla_rs.R")
lib <- file.path("..", "target", "release", "libinla_rs.so")
if (!file.exists(lib)) {
  lib <- file.path("..", "target", "debug", "libinla_rs.so")
}
.inla_rs_dynload(lib)

if (!requireNamespace("Matrix", quietly = TRUE)) {
  stop("Package 'Matrix' is required", call. = FALSE)
}

out_dir <- file.path(script_dir, "spde_validation")
dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)

# 1. Dummy observation coordinates
set.seed(42)
n_points <- 200L
coords <- matrix(runif(n_points * 2L, 0, 10), ncol = 2L)
colnames(coords) <- c("x", "y")

# 2. Triangulated mesh (lattice stand-in for inla.mesh.2d)
pad <- 0.5
mesh <- inla_rs_lattice_mesh(
  xlim = range(coords[, 1]) + c(-pad, pad),
  ylim = range(coords[, 2]) + c(-pad, pad),
  nx = 21L,
  ny = 21L
)
verts <- mesh$vertices
tris <- mesh$triangles

# 3–4. FEM blocks + Matérn SPDE precision (alpha = 2 style: τ²(κ⁴C + 2κ²G + G C⁻¹ G))
fem <- inla_rs_fem_blocks_mesh(verts, tris)
C_matrix <- fem$c0
G_matrix <- fem$g1

kappa <- 1.5
tau <- 1.0
Q <- inla_rs_spde_precision_mesh_csc(verts, tris, kappa = kappa, tau = tau)

cat("Mesh nodes (n):", nrow(verts), "\n")
cat("Number of triangles:", nrow(tris), "\n")
cat("Dimensions of C (Mass):", paste(dim(C_matrix), collapse = " x "), "\n")
cat("Dimensions of G (Stiffness):", paste(dim(G_matrix), collapse = " x "), "\n")
cat("Q nnz:", length(Q@x), "\n")

# Sample x ~ N(0, Q^{-1}) via dense Cholesky (mesh is moderate)
set.seed(42)
Qd <- as.matrix(Q)
Qd <- (Qd + t(Qd)) / 2
# Tiny ridge for numerical SPD (boundary modes can be near-singular)
diag(Qd) <- diag(Qd) + 1e-8
Rchol <- chol(Qd)
z <- rnorm(nrow(Qd))
sample_field <- as.numeric(backsolve(Rchol, z))

# Project sample onto observation locations for a quick fit check
A_obs <- inla_rs_spde_projector_csc(verts, tris, coords[, 1], coords[, 2])
eta_true <- as.numeric(A_obs %*% sample_field)
y_obs <- eta_true + rnorm(n_points, sd = 0.15)

res <- inla_rs_spde(
  y = y_obs,
  loc = coords,
  vertices = verts,
  triangles = tris,
  initial_theta = c(log(tau), log(kappa)),
  obs_precision = 1 / (0.15^2),
  constrain = FALSE
)
cat(
  "SPDE fit mode (log_tau, log_kappa):",
  paste(round(res$mode, 4), collapse = ", "),
  "\n"
)
cat("SPDE mlik:", round(res$marginal_log_lik, 4), "\n")

# ==========================================
# EXPORTING FIGURES
# ==========================================

plot_mesh <- function(verts, tris, main = "Mesh") {
  plot(
    verts,
    type = "n",
    asp = 1,
    xlab = "X",
    ylab = "Y",
    main = main
  )
  for (i in seq_len(nrow(tris))) {
    idx <- c(tris[i, ], tris[i, 1])
    lines(verts[idx, 1], verts[idx, 2], col = "grey40", lwd = 0.4)
  }
}

# Figure 1: Triangulated mesh + observation points
png(file.path(out_dir, "spde_mesh_validation.png"), width = 800, height = 800, res = 120)
plot_mesh(verts, tris, main = "SPDE Triangulated Mesh (rust-inla)")
points(coords, col = "red", pch = 16, cex = 0.55)
dev.off()

# Figure 2: Sparsity pattern of stiffness G
png(file.path(out_dir, "spde_stiffness_sparsity.png"), width = 800, height = 800, res = 120)
Matrix::image(
  G_matrix,
  main = "Sparsity Pattern of Stiffness Matrix (G)",
  sub = "Non-zero elements are highlighted",
  xlab = "column",
  ylab = "row"
)
dev.off()

# Figure 3: Simulated Matérn field on a regular projection grid
gx <- seq(min(verts[, 1]), max(verts[, 1]), length.out = 80L)
gy <- seq(min(verts[, 2]), max(verts[, 2]), length.out = 80L)
grid <- as.matrix(expand.grid(x = gx, y = gy))
A_grid <- inla_rs_spde_projector_csc(verts, tris, grid[, 1], grid[, 2])
field_vec <- as.numeric(A_grid %*% sample_field)
field_grid <- matrix(field_vec, nrow = length(gx), ncol = length(gy))

png(file.path(out_dir, "spde_simulated_field.png"), width = 800, height = 800, res = 120)
image(
  gx,
  gy,
  field_grid,
  col = terrain.colors(100),
  main = "Simulated Matérn Spatial Field (rust-inla)",
  xlab = "X",
  ylab = "Y",
  asp = 1
)
contour(gx, gy, field_grid, add = TRUE, col = "black", labcex = 0.6)
dev.off()

# Figure 4: Posterior mean field from SPDE fit
field_fit <- matrix(
  as.numeric(A_grid %*% res$latent_means),
  nrow = length(gx),
  ncol = length(gy)
)
png(file.path(out_dir, "spde_fitted_field.png"), width = 800, height = 800, res = 120)
image(
  gx,
  gy,
  field_fit,
  col = terrain.colors(100),
  main = "Fitted SPDE Posterior Mean Field",
  xlab = "X",
  ylab = "Y",
  asp = 1
)
points(coords, pch = 16, cex = 0.35, col = adjustcolor("black", 0.35))
dev.off()

cat("Figures exported to:", normalizePath(out_dir), "\n")

File: py-inla/validate_spde.py
#!/usr/bin/env python3
"""SPDE mesh / FEM / field demo using rust-inla only (no classic R-INLA).

Mirrors r-inla/validate_spde.R:
  coords → lattice mesh → FEM (C, G) → Q → sample field → fit → figures
"""

from __future__ import annotations

import sys
from pathlib import Path

import numpy as np

# Allow running before install if PYTHONPATH includes python/
HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE / "python"))

from inla import _native as core  # noqa: E402
from inla.spde import (  # noqa: E402
    fem_blocks_mesh,
    lattice_mesh,
    precision_matrix,
    projector_matrix,
)

try:
    import matplotlib

    matplotlib.use("Agg")
    import matplotlib.pyplot as plt
except ImportError as exc:
    raise SystemExit(
        "matplotlib is required for figure export; install with: pip install matplotlib"
    ) from exc


def main() -> None:
    out_dir = HERE / "spde_validation"
    out_dir.mkdir(parents=True, exist_ok=True)

    rng = np.random.default_rng(42)
    n_points = 200
    coords = rng.uniform(0.0, 10.0, size=(n_points, 2))

    pad = 0.5
    mesh = lattice_mesh(
        xlim=(coords[:, 0].min() - pad, coords[:, 0].max() + pad),
        ylim=(coords[:, 1].min() - pad, coords[:, 1].max() + pad),
        nx=21,
        ny=21,
    )
    verts = mesh["vertices"]
    tris = mesh["triangles"]
    vert_tuples = [(float(x), float(y)) for x, y in verts]
    tri_tuples = [(int(a), int(b), int(c)) for a, b, c in tris]

    fem = fem_blocks_mesh(verts, tris)
    # Keep PyCscMatrix alive: to_scipy() is a view into Rust buffers.
    c_py, g_py = fem["c0"], fem["g1"]
    c_matrix = c_py.to_scipy().tocsc().copy()
    g_matrix = g_py.to_scipy().tocsc().copy()

    kappa = 1.5
    tau = 1.0
    q_py = precision_matrix(verts, tris, kappa=kappa, tau=tau)
    q = q_py.to_scipy().tocsc().copy()

    print(f"Mesh nodes (n): {verts.shape[0]}")
    print(f"Number of triangles: {tris.shape[0]}")
    print(f"Dimensions of C (Mass): {c_matrix.shape[0]} x {c_matrix.shape[1]}")
    print(f"Dimensions of G (Stiffness): {g_matrix.shape[0]} x {g_matrix.shape[1]}")
    print(f"Q nnz: {q.nnz}")

    # Sample x ~ N(0, Q^{-1}) via dense Cholesky
    qd = q.toarray()
    qd = 0.5 * (qd + qd.T)
    qd.flat[:: qd.shape[0] + 1] += 1e-8
    rchol = np.linalg.cholesky(qd)
    sample_field = np.linalg.solve(rchol.T, rng.standard_normal(qd.shape[0]))

    a_obs_py = projector_matrix(verts, tris, coords[:, 0], coords[:, 1])
    a_obs = a_obs_py.to_scipy().tocsc().copy()
    eta_true = a_obs @ sample_field
    y_obs = eta_true + rng.normal(0.0, 0.15, size=n_points)

    def build_prior(theta):
        t = float(np.exp(theta[0]))
        k = float(np.exp(theta[1]))
        return core.spde_precision_matrix(vert_tuples, tri_tuples, kappa=k, tau=t)

    def log_prior(theta):
        return float(-0.05 * (theta[0] ** 2 + theta[1] ** 2))

    obs = [{"family": "gaussian", "y": float(yi), "precision": 1.0 / (0.15**2)} for yi in y_obs]
    a_py = core.spde_projector_matrix(
        vert_tuples, tri_tuples, coords[:, 0].tolist(), coords[:, 1].tolist()
    )
    res = core.run_inla_inference(
        [np.log(tau), np.log(kappa)],
        build_prior,
        log_prior,
        obs,
        strategy="ccd",
        a=a_py,
    )
    print(
        "SPDE fit mode (log_tau, log_kappa):",
        ", ".join(f"{m:.4f}" for m in res.mode),
    )
    print(f"SPDE mlik: {res.marginal_log_lik:.4f}")

    # --- figures ---
    fig, ax = plt.subplots(figsize=(8, 8), dpi=120)
    for a, b, c in tris:
        idx = [a, b, c, a]
        ax.plot(verts[idx, 0], verts[idx, 1], color="0.4", lw=0.4)
    ax.scatter(coords[:, 0], coords[:, 1], c="red", s=8, zorder=3)
    ax.set_aspect("equal")
    ax.set_title("SPDE Triangulated Mesh (rust-inla)")
    ax.set_xlabel("X")
    ax.set_ylabel("Y")
    fig.tight_layout()
    fig.savefig(out_dir / "spde_mesh_validation.png")
    plt.close(fig)

    fig, ax = plt.subplots(figsize=(8, 8), dpi=120)
    ax.spy(g_matrix, markersize=0.4, color="black")
    ax.set_title("Sparsity Pattern of Stiffness Matrix (G)")
    ax.set_xlabel("column")
    ax.set_ylabel("row")
    fig.tight_layout()
    fig.savefig(out_dir / "spde_stiffness_sparsity.png")
    plt.close(fig)

    gx = np.linspace(verts[:, 0].min(), verts[:, 0].max(), 80)
    gy = np.linspace(verts[:, 1].min(), verts[:, 1].max(), 80)
    grid_x, grid_y = np.meshgrid(gx, gy, indexing="ij")
    a_grid_py = projector_matrix(verts, tris, grid_x.ravel(), grid_y.ravel())
    a_grid = a_grid_py.to_scipy().tocsc().copy()
    field_grid = (a_grid @ sample_field).reshape(grid_x.shape)

    fig, ax = plt.subplots(figsize=(8, 8), dpi=120)
    im = ax.imshow(
        field_grid.T,
        origin="lower",
        extent=(gx.min(), gx.max(), gy.min(), gy.max()),
        cmap="terrain",
        aspect="equal",
    )
    ax.contour(gx, gy, field_grid.T, colors="black", linewidths=0.5)
    fig.colorbar(im, ax=ax, fraction=0.046, pad=0.04)
    ax.set_title("Simulated Matérn Spatial Field (rust-inla)")
    ax.set_xlabel("X")
    ax.set_ylabel("Y")
    fig.tight_layout()
    fig.savefig(out_dir / "spde_simulated_field.png")
    plt.close(fig)

    field_fit = (a_grid @ np.asarray(res.latent_means, dtype=float)).reshape(grid_x.shape)
    fig, ax = plt.subplots(figsize=(8, 8), dpi=120)
    im = ax.imshow(
        field_fit.T,
        origin="lower",
        extent=(gx.min(), gx.max(), gy.min(), gy.max()),
        cmap="terrain",
        aspect="equal",
    )
    ax.scatter(coords[:, 0], coords[:, 1], c="k", s=4, alpha=0.35)
    fig.colorbar(im, ax=ax, fraction=0.046, pad=0.04)
    ax.set_title("Fitted SPDE Posterior Mean Field")
    ax.set_xlabel("X")
    ax.set_ylabel("Y")
    fig.tight_layout()
    fig.savefig(out_dir / "spde_fitted_field.png")
    plt.close(fig)

    print(f"Figures exported to: {out_dir.resolve()}")


if __name__ == "__main__":
    main()
